src/cluster_import.c

```diff
@@ -0,0 +1,1915 @@
+            updateSlotExportIfOwnershipChanged(link);
+            return;
+
+        /* Terminal states */
+        case SLOT_MIGRATION_LINK_SUCCESS:
+        case SLOT_MIGRATION_LINK_CANCELLED:
+        case SLOT_MIGRATION_LINK_FAILED:
+            /* Nothing to do right now in cron */
+            return;
+        }
+    }
+}
+
+void resetSlotMigrationLink(slotMigrationLink *link) {
+    /* Only one of client or conn should be set. */
+    serverAssert(!link->client || !link->conn);
+    if (link->client) {
+        if (!link->client->flag.close_after_reply) {
+            freeClientAsync(link->client);
+        }
+        link->client = NULL;
+    } else if (link->conn) {
+        connClose(link->conn);
+        link->conn = NULL;
+    }
+}
+
+void freeSlotMigrationLink(void *o) {
+    slotMigrationLink *link = o;
+    resetSlotMigrationLink(link);
+    listRelease(link->slot_ranges);
+    sdsfree(link->slot_ranges_str);
+    if (link->status_msg) {
+        sdsfree(link->status_msg);
+    }
+    zfree(o);
+}
+
+void initClusterSlotMigrationLinkList(void) {
+    server.cluster->slot_migration_links = listCreate();
+    listSetFreeMethod(server.cluster->slot_migration_links, freeSlotMigrationLink);
+}
+
+int shouldCleanupSlotMigrationLink(slotMigrationLink *link, unsigned long idx) {
+    return (link->state == SLOT_MIGRATION_LINK_CANCELLED ||
+            link->state == SLOT_MIGRATION_LINK_FAILED ||
+            link->state == SLOT_MIGRATION_LINK_SUCCESS) &&
+           ((server.cluster_slot_migration_log_ttl &&
+             server.unixtime - link->last_update >= server.cluster_slot_migration_log_ttl) ||
+            idx >= server.cluster_slot_migration_log_max_len);
+}
+
+const char *slotMigrationLinkStateToString(slotMigrationLinkState state) {
+    switch (state) {
+    case SLOT_IMPORT_RECONNECT: return "reconnecting";
+    case SLOT_IMPORT_CONNECTING: return "connecting";
+    case SLOT_IMPORT_AUTHENTICATING: return "authenticating";
+    case SLOT_IMPORT_START_SNAPSHOT:
+    case SLOT_IMPORT_RECEIVE_SNAPSHOT: return "snapshotting";
+    case SLOT_IMPORT_RECEIVE_STREAM: return "replicating";
+    case SLOT_IMPORT_FAILOVER_WAITING_FOR_PAUSED: return "failover-syncing";
+    case SLOT_IMPORT_FAILOVER_REQUESTED: return "failover-requested";
+    case SLOT_IMPORT_FAILOVER_GRANTED: return "failover-granted";
+    case SLOT_IMPORT_FINISHED_WAITING_TO_CLEANUP: return "cleaning-up";
+
+    case SLOT_EXPORT_WAITING_TO_SNAPSHOT: return "waiting-to-snapshot";
+    case SLOT_EXPORT_SNAPSHOTTING: return "snapshotting";
+    case SLOT_EXPORT_STREAMING: return "replicating";
+    case SLOT_EXPORT_FAILOVER_PAUSED: return "failover-paused";
+    case SLOT_EXPORT_FAILOVER_GRANTED: return "failover-granted";
+
+    case SLOT_MIGRATION_LINK_SUCCESS: return "success";
+    case SLOT_MIGRATION_LINK_CANCELLED: return "cancelled";
+    case SLOT_MIGRATION_LINK_FAILED: return "failed";
+    default: return "unknown";
```

Comment:
```suggestion
```
Remove the default and return unknown outside of it. If we ever add a new state, we'll get a warning that this case switch is missing a state.